metricserver/metricserver.go

```diff
@@ -0,0 +1,66 @@
+package metricserver
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/otel/exporters/prometheus"
+	"go.opentelemetry.io/otel/metric"
+	sdk "go.opentelemetry.io/otel/sdk/metric"
+)
+
+const (
+	DEFAULT_PORT = 2112
+)
+
+type MetricServerConfiguration func(*MetricServer) error
+
+type MetricServer struct {
+	port int
+
+	meterProvider *sdk.MeterProvider
+}
+
+func New(cfgs ...MetricServerConfiguration) (*MetricServer, error) {
```

Comment:
Do we need to implement a shutdown function to gracefully shutdown the metric server in the `fluent-bit-kinesis.go`? Seems that the MetricProvider supports this: https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#MeterProvider.Shutdown

Reply:
I thought about this too. 

From the description of `Shutdown` it looks like it just flushes and releases any computational resources.

For our implementation, we use a pull-based approach using a Prometheus exporter. With this, we aren't buffering any metrics to send to an external agent.

IIUC, if the main thread gets terminated, MetricProvider and Prometheus Server should be killed along with it, which I may be wrong but would also release computational resources? 

WDYT?
Happy to implement a shutdown procedure for MeterProvider.

Reply:
hmm you're right. in that case we may want to have a graceful shutdown just for the http server. Meaning if the `/metric` endpoint is being called and we receive a shutdown signal while we are pulling , it will wait for the http server to gracefully shutdown. However, not sure how important it is to not loosing these metrics. Maybe it does not worth the effort and we can afford to loose them

Reply:
IIUC, there is no blocking thread as `main()` is not executed. I believe the go functions are called as C bindings so doing any grace shutdown would be tricky. WDYT?

Reply:
It seems that Go already has taken care of the complexities:
https://pkg.go.dev/os/signal#hdr-Non_Go_programs_that_call_Go_code

Reply:
I just pushed a shutdown function and it would be invoked in `FLBPluginExit`, this function is exported and will be called when shutdown occurs. This implementation isn't the cleanest approach.

Testing this in telemetry-test when killing the pod gives:
```
time="2023-06-20T03:44:19Z" level=info msg="shutting down metrics server"
time="2023-06-20T03:44:19Z" level=fatal msg="http: Server closed"
```

Reply:
I added a Shutdown procedure in `FLBPluginExit` which is called when exiting. I changed how the Server starts.

The setup isn't ideal and clean but it works. WDYT?

```
time="2023-06-20T03:54:34Z" level=info msg="shutting down metrics server"
time="2023-06-20T03:54:34Z" level=info msg="http: Server closed"
time="2023-06-20T03:54:34Z" level=info msg="shutdown metrics server complete"
```

Reply:
yeap, makes sense to leave it to FluentBit to handle the signal and close the plugins